prime-router/docs/design/hl7-validation/design.md

```diff
@@ -0,0 +1,115 @@
+# Problem Statement
+As a sender/receiver I want submitted/received data to be validated to ensure data integrity and quality.
+
+There needs to exist a collection of settings and reusable components to perform varying types of validation at various points in the HL7 message's lifecycle.
+
+## Proposed Design
+![hl7-validation-annotated-architecture-diagram.png](annotated-architecture-diagram.png)
+
+We propose two validation "checkpoints":
+1. Sender validation
+2. Receiver Validation
+
+This will use the setting proposed in the FHIR validation design doc called `validationProfile`. This is reusable because
+senders and receivers can only be set up to receive one format at a time, so there should not be confusion over which 
+validationProfile pertains to which format. 
+
+Each validation checkpoint shall run its configured validation and in the event of a failure: 
+- actionLog the result and error information
+- notify the sender
+
+## Background Information/Dev Notes
+### The HAPI HL7 library
+The HAPI library offers 3 "levels" of message validation:
+1. Basic Message Validation - very basic validation on message length and some segments, but the article
+(https://saravanansubramanian.com/hl72xhapivalidation/) said this will not be enough
+2. Custom Message Validation - Extends basic and enabled through the extensibility offered by the message validation 
+framework itself. Allows you to override default behaviour.
+3. Validation by the use of Conformance Profiles - enable the sharing of any stringent message specifications involving 
+specific workflows to an open registry that everyone can use. They push this one. We may be able to download something
+that others have created to use as a basis for our custom message validation. This is worth looking into. We need to 
+determine if it would be appropriate for us to contribute to this community.
+
+Behavior that can be controlled:
+- Whether the message validation should be activated or not (it is turned on by default)
+- How to deal with unrecognized message versions, segments, fields, etc. during the parsing process
+- Behavior around how the parser should encode segments and fields even if no data content is present
+
+Key Classes:
+- `ValidationContext` - manages a collection of codified rules for handling messages, fields, and primitive types
+- `ValidationRuleBuilder` - what you use to create the validation rules
+- `MessageVisitor` - the interface the classes implement when they are in charge of doing the actual comparison
+- `ValidationException` - thrown when validation fails
+- `ValidationExceptionHandlerFactory` - interface that you can implement in order to control the exception handling
+ 
+<details>
```

Comment:
This chunk should be in a markdown code block with a description of what it is doing.